tests/test_recombination_map.py

```diff
@@ -0,0 +1,437 @@
+
+    def _physical_to_genetic_zero_rate(self, x):
+        """
+        If we have a zero recombination rate throughout, then we only have
+        two possible values. Any value < L maps to 0, as this is the start
+        of the interval. If x = L, then we map to num_loci because the interval
+        is half-open. This is consistent with the behaviour we get when we
+        have an interval of zero recombination within the map: any value
+        within the interval is mapped down to the start coordinate, and the
+        endpoint signifies the beginning of the next interval.
+        """
+        ret = 0
+        if x >= self._sequence_length:
+            ret = self._num_loci
+        return ret
+
+    def _genetic_to_physical_zero_rate(self, v):
+        """
+        If we have a zero recombination rate throughout then 0 maps to the
+        start of the interval and everything else maps to L.
+        """
+        return 0 if v == 0 else self._sequence_length
+
+    def physical_to_genetic(self, x):
+        if self.get_total_recombination_rate() == 0:
+            # TODO this check at zero shouldn't be necessary as should be
+            # able to use the same code to deal with an interval within
+            # the map.
+            return self._physical_to_genetic_zero_rate(x)
+        s = 0
+        last_phys_x = 0
+        j = 1
+        while j < len(self._positions) - 1 and x > self._positions[j]:
+            phys_x = self._positions[j]
+            rate = self._rates[j - 1]
+            s += (phys_x - last_phys_x) * rate
+            j += 1
+            last_phys_x = phys_x
+        rate = self._rates[j - 1]
+        s += (x - last_phys_x) * rate
+        ret = s / self.get_total_recombination_rate()
+        return ret * self._num_loci
+
+    def physical_to_discrete_genetic(self, x):
+        return int(round(self.physical_to_genetic(x)))
+
+    def genetic_to_physical(self, v):
+        if self.get_total_recombination_rate() == 0:
+            # TODO this check at zero shouldn't be necessary as should be
+            # able to use the same code to deal with an interval within
+            # the map.
+            return self._genetic_to_physical_zero_rate(v)
+        # v is expressed in [0, m]. Rescale it back into the range
+        # (0, total_mass).
+        u = (v / self._num_loci) * self.get_total_recombination_rate()
+        s = 0
+        last_phys_x = 0
+        rate = self._rates[0]
+        j = 1
+        while j < len(self._positions) and s < u:
+            phys_x = self._positions[j]
+            rate = self._rates[j - 1]
+            s += (phys_x - last_phys_x) * rate
+            j += 1
+            last_phys_x = phys_x
+        if rate == 0:
+            # If the rate for the current interval is zero, map back to
+            # first position unless we're at the end of the interval
+            assert False, "this hasn't been tested"
+            y = last_phys_x if s < u else self._positions[j]
```

Comment:
This isn't the first position. I think we'd want `y = self._positions[j-1]` if we knew there was only one region of zero recombination; but in the presence of more than one consecutive one this won't be right.

Reply:
also note that as written, `last_phys_x` and `self._positions[j]` are the same.